File: src/domain/repositories/auth_repository.rs
use anyhow::Result;
use async_trait::async_trait;

use crate::domain::models::user::{AuthToken, User};

/// Repository interface for authentication
#[async_trait]
pub trait AuthRepository: Send + Sync {
    /// Login to the server
    async fn login(&self, server_url: &str, username: &str, password: &str) -> Result<(AuthToken, User)>;
    
    /// Refresh the authentication token
    async fn refresh_token(&self, refresh_token: &str) -> Result<AuthToken>;
    
    /// Logout from the server
    async fn logout(&self) -> Result<()>;
    
    /// Get the current user profile
    async fn get_current_user(&self) -> Result<User>;
    
    /// Store the authentication token securely
    async fn store_token(&self, token: &AuthToken) -> Result<()>;
    
    /// Retrieve the stored authentication token
    async fn get_stored_token(&self) -> Result<Option<AuthToken>>;
    
    /// Clear the stored authentication token
    async fn clear_stored_token(&self) -> Result<()>;
}

File: lib/domain/repositories/auth_repository.dart
import '../entities/auth_tokens_entity.dart';
import '../entities/user_entity.dart';

abstract class AuthRepository {
  /// Check if system has an admin configured.
  Future<({bool adminExists, bool registrationEnabled})> getStatus();

  /// Initial admin setup.
  Future<(UserEntity, AuthTokens)?> setup({
    required String username,
    required String email,
    required String password,
  });

  /// Login with username and password.
  Future<(UserEntity, AuthTokens)> login({
    required String username,
    required String password,
  });

  /// Register a new user.
  Future<(UserEntity, AuthTokens)> register({
    required String username,
    required String email,
    required String password,
  });

  /// Refresh the access token.
  Future<AuthTokens> refreshToken(String refreshToken);

  /// Get the current user profile.
  Future<UserEntity> getCurrentUser();

  /// Change password.
  Future<void> changePassword({
    required String currentPassword,
    required String newPassword,
  });

  /// Logout (invalidate session server-side).
  Future<void> logout();
}
